### PR TITLE
fix(vm):  lifecycle vm

### PR DESCRIPTION
**Pull request by yaroslavborbat**
## Description
fix secret keys for provisioning. [userData, userdata]
add observedGeneration to VirtualMachine.Status
remove OwnerReference from pvc when DataVolume is removed

## Checklist
- [ ] The code is covered by unit tests.
- [ ] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [X] Changes were tested in the Kubernetes cluster manually.

### CODE REVIEW

**Review by LopatinDmitr — DISMISSED** [review]

File: images/virtualization-artifact/pkg/controller/vm/internal/lifecycle.go [hunk 1]
```diff
@@ -58,6 +58,22 @@ func (h *LifeCycleHandler) Handle(ctx context.Context, s state.VirtualMachineSta
...
 	changed := s.VirtualMachine().Changed()
 
+	defer func() {
+		if changed == nil {
+			return
+		}
+		if len(changed.Status.Conditions) == 0 {
+			changed.Status.ObservedGeneration = changed.GetGeneration()
+			return
+		}
```
Comment (diafour) [inline images/virtualization-artifact/pkg/controller/vm/internal/lifecycle.go:68]: Shouldn't we always update ObservedGeneration on every reconcile despite of changes?

File: images/virtualization-artifact/pkg/controller/vm/internal/provisioning.go [hunk 2]
```diff
@@ -131,10 +131,18 @@ func (h *ProvisioningHandler) genConditionFromSecret(ctx context.Context, builde
...
+		if _, ok := secret.Data[key]; ok {
+			found = true
+			break
+		}
+	}
+	if !found {
 		builder.Status(metav1.ConditionFalse).
 			Reason2(vmcondition.ReasonProvisioningNotReady).
-			Message(fmt.Sprintf("Secret %q doesn't have key \"userdata\".", secretKey.String()))
+			Message(fmt.Sprintf("Secret %q doesn't have any keys such as %v.", keys, secretKey.String()))
```
Comment (diafour) [inline images/virtualization-artifact/pkg/controller/vm/internal/provisioning.go:145]:
```suggestion
			Message(fmt.Sprintf("Secret %q should has one of data fields %v.", keys, secretKey.String()))
```

File: images/virtualization-artifact/pkg/controller/service/disk_service.go [hunk 3]
```diff
@@ -148,6 +149,17 @@ func (s DiskService) CleanUpSupplements(ctx context.Context, sup *supplements.Ge
 		if err != nil && !k8serrors.IsNotFound(err) {
 			return false, err
 		}
+		pvc, err := s.GetPersistentVolumeClaim(ctx, sup)
```
Comment (Isteb4k) [inline images/virtualization-artifact/pkg/controller/service/disk_service.go:152]: Check pvc == nil

**Review by Isteb4k — CHANGES_REQUESTED** [review]

File: images/virtualization-artifact/pkg/controller/service/disk_service.go (hunk 3 above)
Comment (Isteb4k) [reply images/virtualization-artifact/pkg/controller/service/disk_service.go:152]: by the way, the DataVolume can be deleted by the hands of users. In this case, the ownerRef will still be in the in PVC

File: images/virtualization-artifact/pkg/controller/vm/internal/lifecycle.go (hunk 1 above)
Comment (yaroslavborbat) [reply images/virtualization-artifact/pkg/controller/vm/internal/lifecycle.go:68]: I think we should update the ObservedGeneration when the ObservedGeneration of all conditions has been updated

**Review by Isteb4k — APPROVED** [review]

**Pull request merged**